inference/generativeai/llm-workshop/lab4-openchatkit/deploy_openchatkit_on_sagemaker_deepspeed.ipynb

```diff
@@ -0,0 +1,1141 @@
+    "    model_location = properties[\"model_dir\"]\n",
+    "    if \"model_id\" in properties:\n",
+    "        model_location = properties[\"model_id\"]\n",
+    "    logging.info(f\"Loading model in {model_location}\")\n",
+    "\n",
+    "    tokenizer = AutoTokenizer.from_pretrained(model_location)\n",
+    "    dtype = torch.float16\n",
+    "\n",
+    "    model = AutoModelForCausalLM.from_pretrained(\n",
+    "        model_location, low_cpu_mem_usage=True, torch_dtype=dtype\n",
+    "    )\n",
+    "\n",
+    "    if \"dtype\" in properties:\n",
+    "        if properties[\"dtype\"] == \"float16\":\n",
+    "            model.to(torch.float16)\n",
+    "        if properties[\"dtype\"] == \"bfloat16\":\n",
+    "            model.to(torch.bfloat16)\n",
+    "\n",
+    "    logging.info(f\"Starting DeepSpeed init with TP={tensor_parallel}\")\n",
+    "    model = deepspeed.init_inference(\n",
+    "        model,\n",
+    "        mp_size=tensor_parallel,\n",
+    "        dtype=model.dtype,\n",
+    "        replace_method=\"auto\",\n",
+    "        replace_with_kernel_inject=True,\n",
+    "    )\n",
+    "    return model.module, tokenizer\n",
+    "\n",
+    "\n",
+    "model = None\n",
+    "tokenizer = None\n",
+    "generator = None\n",
+    "\n",
+    "\n",
+    "def run_inference(model, tokenizer, data, params):\n",
+    "    generate_kwargs = params\n",
+    "    tokenizer.pad_token = tokenizer.eos_token\n",
+    "    input_tokens = tokenizer.batch_encode_plus(data, return_tensors=\"pt\", padding=True)\n",
+    "    for t in input_tokens:\n",
+    "        if torch.is_tensor(input_tokens[t]):\n",
+    "            input_tokens[t] = input_tokens[t].to(torch.cuda.current_device())\n",
+    "    outputs = model.generate(**input_tokens, **generate_kwargs)\n",
+    "    return tokenizer.batch_decode(outputs, skip_special_tokens=True)\n",
+    "\n",
+    "\n",
+    "def handle(inputs: Input):\n",
+    "    global model, tokenizer\n",
+    "    if not model:\n",
+    "        model, tokenizer = load_model(inputs.get_properties())\n",
+    "\n",
+    "    if inputs.is_empty():\n",
+    "        # Model server makes an empty call to warmup the model on startup\n",
+    "        return None\n",
+    "    data = inputs.get_as_json()\n",
+    "\n",
+    "    input_sentences = data[\"inputs\"]\n",
+    "    params = data[\"parameters\"]\n",
+    "\n",
+    "    outputs = run_inference(model, tokenizer, input_sentences, params)\n",
+    "    result = {\"outputs\": outputs}\n",
+    "    return Output().add_as_json(result)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "5dd60a29",
+   "metadata": {},
+   "source": [
+    "**Image URI for the DJL container is being used here**"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "3884d357",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "inference_image_uri = (\n",
```

Comment:
use urlretrive:

https://docs.djl.ai/docs/demos/aws/sagemaker/large-model-inference/standard_template_with_LMI_solution.html#getting-the-container-image-uri

Reply:
Done